File: python/eos/cubic_eos_base.py

from scipy.constants import gas_constant
import numpy as np

class CubicEosBase:
    """
    Base class for cubic equation of states.
    """

    def __init__(self, pc, tc, omega_a, omega_b):
        """
        Parameters
        ----------
        pc : float
            Critical pressure
        tc : float
            Critical temperature
        omega_a : float
            Coefficient for attraction parameter
        omega_b : float
            Coefficient for repulsion parameter
        """
        self.__pc = pc
        self.__tc = tc
        self.__omega_a = omega_a
        self.__omega_b = omega_b
        self.__a = self._calc_attraction_param()
        self.__b = self._calc_repulsion_param()

    @property
    def critical_pressure(self):
        return self.__pc

    @property
    def critical_temperature(self):
        return self.__tc

    @property
    def attraction_param(self):
        return self.__a

    @property
    def repulsion_param(self):
        return self.__b

    def reduced_pressure(self, p):
        """
        Computes reduced pressure

        Parameters
        ----------
        p : float
            Pressure

        Returns
        -------
        float
            Reduced pressure
        """
        return p/self.__pc

    def reduced_temperature(self, t):
        """
        Computes reduced temperature

        Parameters
        ----------
        t : float
            Temperature

        Returns
        -------
        float
            Reduced temperature
        """
        return t/self.__tc

    def _calc_attraction_param(self):
        """ Computes attraction parameter """
        return self.__omega_a*(gas_constant*self.__tc)**2/self.__pc

    def _calc_repulsion_param(self):
        """ Computes repulsion parameter """
        return self.__omega_b*gas_constant*self.__tc/self.__pc

    def _calc_reduced_attraction_param(self, p, t, alpha=1.0):
        """
        Computes A = alpha*a*P/(R*T)^2

        Parameters
        ----------
        p : float
            Pressure
        t : float
            Temperature
        alpha : float
            Temperature correction factor. The default value is 1.0.

        Returns
        -------
        float
            A = alpha*a*P/(R*T)^2
        """
        pr = p/self.__pc
        tr = t/self.__tc
        return self.__omega_a*alpha*pr/tr**2

    def _calc_reduced_repulsion_param(self, p, t):
        """
        Computes B = b*P/(R*T)

        Parameters
        ----------
        p : float
            Pressure
        t : float
            Temperature

        Returns
        -------
        float
            B = b*P/(R*T)
        """
        pr = p/self.__pc
        tr = t/self.__tc
        return self.__omega_b*pr/tr

    @staticmethod
    def solve_cubic_eq(eq):
        """
        Solves a cubic equation

        Parameters
        ----------
        eq : list
            A list of coefficients of a cubic equation
        
        Returns
        -------
        numpy.array
            Real roots of a cubic equation in the ascending order
        """
        x = np.roots(eq)
        return np.sort(np.array([xi.real for xi in x if not np.iscomplex(xi)]))

File: matlab/+eos/+purecomp/CubicEosBase.m
classdef CubicEosBase
    % Base class for two-parameter cubic equation of state.
    
    properties (SetAccess = private)
        CriticalPressure    % Critical pressure [Pa]
        CriticalTemperature % Critical temperature [K]
        MolecularWeight     % Molecular weight [g/mol]
        OmegaA              % Coefficient for attraction parameter
        OmegaB              % Coefficient for repulsion parameter
        AttractionParam     % Attraction parameter
        RepulsionParam      % Repulsion parameter
    end
    methods
        function obj = CubicEosBase(OmegaA,OmegaB,Pc,Tc,Mw)
            % Construct cubic EOS
            %
            % Parameters
            % ----------
            % OmegaA : Coefficient for attraction parameter
            % OmegaB : Coefficient for repulsion parameter
            % Pc : Critical pressure [Pa]
            % Tc : Critical temperature [K]
            % Mw : Molecular weight [g/mol]
            arguments
                OmegaA (1,1) {mustBeNumeric}
                OmegaB (1,1) {mustBeNumeric}
                Pc (1,1) {mustBeNumeric}
                Tc (1,1) {mustBeNumeric}
                Mw (1,1) {mustBeNumeric}
            end
            obj.OmegaA = OmegaA;
            obj.OmegaB = OmegaB;
            obj.CriticalPressure = Pc;
            obj.CriticalTemperature = Tc;
            obj.MolecularWeight = Mw;
            R = eos.ThermodynamicConstants.Gas;
            obj.AttractionParam = OmegaA*(R*Tc)^2/Pc;
            obj.RepulsionParam = OmegaB*R*Tc/Pc;
        end
        function obj = setParams(obj,Pc,Tc,Mw)
            % Set parameters
            %
            % Parameters
            % ----------
            % Pc : Critical pressure [Pa]
            % Tc : Critical temperature [K]
            % Mw : Molecular weight [g/mol]
            arguments
                obj
                Pc (1,1) {mustBeNumeric}
                Tc (1,1) {mustBeNumeric}
                Mw (1,1) {mustBeNumeric}
            end
            obj.CriticalPressure = Pc;
            obj.CriticalTemperature = Tc;
            obj.MolecularWeight = Mw;
            R = eos.ThermodynamicConstants.Gas;
            obj.AttractionParam = obj.OmegaA*(R*Tc)^2/Pc;
            obj.RepulsionParam = obj.OmegaB*R*Tc/Pc;
        end
        function Pr = reducedPressure(obj,P)
            % Compute reduced pressure
            %
            % Parameters
            % ----------
            % P : Pressure [Pa]
            %
            % Returns
            % -------
            % Pr : Reduced pressure
            Pr = P./obj.CriticalPressure;
        end
        function Tr = reducedTemperature(obj,T)
            % Compute reduced temperature
            %
            % Parameters
            % ----------
            % T : Temperature [K]
            %
            % Returns
            % -------
            % Tr : Reduced temperature
            Tr = T./obj.CriticalTemperature;
        end
        function rho = massDensity(obj,P,T,z)
            % Calculate mass density
            %
            % Parameters
            % ----------
            % P : Pressure [Pa]
            % T : Temperature [K]
            % z : Z-factor
            %
            % Returns
            % -------
            % rho : Mass density [kg/m3]
            R = eos.ThermodynamicConstants.Gas;
            Mw = obj.MolecularWeight;
            rho = P.*Mw*1e-3./(z*R.*T);
        end
        function A = reducedAttractionParam(obj,Pr,Tr,alpha)
            % Compute reduced attraction parameter
            %
            % Parameters
            % ----------
            % Pr : Reduced pressure
            % Tr : Reduced temperature
            % alpha : Temperature correction factor
            %
            % Returns
            % -------
            % A : Reduced attraction parameter
            A = obj.OmegaA*alpha.*Pr./Tr.^2;
        end
        function B = reducedRepulsionParam(obj,Pr,Tr)
            % Compute reduced repulsion parameter
            %
            % Parameters
            % ----------
            % Pr : Reduced pressure
            % Tr : Reduced temperature
            %
            % Returns
            % -------
            % B : Reduced repulsion parameter
            B = obj.OmegaB*Pr./Tr;
        end
        function P = pressure(obj,T,V)
            % Compute pressure
            %
            % P = obj.PRESSURE(T,V)
            %
            % Parameters
            % ----------
            % T : Temperature [K]
            % V : Volume [m3]
            %
            % Returns
            % -------
            % P : Pressure [Pa]
            arguments
                obj {mustBeA(obj,'eos.purecomp.CubicEosBase')}
                T (:,:) {mustBeNumeric}
                V (:,:) {mustBeNumeric}
            end
            Tr = obj.reducedTemperature(T);
            alpha = obj.temperatureCorrectionFactor(Tr);
            a = alpha*obj.AttractionParam;
            b = obj.RepulsionParam;
            P = obj.pressureImpl(T,V,a,b);
        end
        function lnPhi = lnFugacityCoeff(obj,z,s)
            % Compute fugacity coefficients
            %
            % Parameters
            % ----------
            % z : Z-factors
            % s : struct containing parameters
            %
            % Returns
            % -------
            % lnPhi : Fugacity coefficients
            arguments
                obj {mustBeA(obj,'eos.purecomp.CubicEosBase')}
                z (:,1) {mustBeNumeric}
                s struct
            end
            lnPhi = exp(obj.lnFugacityCoeffImpl(z,s.A,s.B));
        end
        function phi = fugacityCoeff(obj,z,s)
            % Compute fugacity coefficients
            %
            % Parameters
            % ----------
            % z : Z-factors
            % s : struct containing parameters
            %
            % Returns
            % -------
            % phi : Fugacity coefficients
            arguments
                obj {mustBeA(obj,'eos.purecomp.CubicEosBase')}
                z (:,1) {mustBeNumeric}
                s struct
            end
            phi = exp(obj.lnFugacityCoeffImpl(z,s.A,s.B));
        end
        function [z,s] = zFactors(obj,P,T)
            % Compute Z-factors
            %
            % Parameters
            % ----------
            % P : Pressure [Pa]
            % T : Temperature [K]
            %
            % Returns
            % -------
            % z : Z-factors
            % s : struct containing parameters
            arguments
                obj {mustBeA(obj,'eos.purecomp.CubicEosBase')}
                P (1,1) {mustBeNumeric}
                T (1,1) {mustBeNumeric}
            end
            Pr = obj.reducedPressure(P);
            Tr = obj.reducedTemperature(T);
            alpha = obj.temperatureCorrectionFactor(Tr);
            A = obj.reducedAttractionParam(Pr,Tr,alpha);
            B = obj.reducedRepulsionParam(Pr,Tr);
            x = roots(obj.zFactorCubicEq(A,B));
            z = x(imag(x) == 0);
            if nargout > 1
                s.P = P;
                s.T = T;
                s.A = A;
                s.B = B;
            end
        end
        function P = tripleRootPressureRange(obj,T)
            % Compute pressure range with triple roots of Z-factors at a
            % given temperature
            %
            % Parameters
            % ----------
            % T : Temperature [K]
            %
            % Returns
            % -------
            % P : Pressures [Pa]
            arguments
                obj {mustBeA(obj,'eos.purecomp.CubicEosBase')}
                T (1,1) {mustBeNumeric}
            end
            if T >= obj.CriticalTemperature
                error("Error. \nT %f must be less than Tc %f.", ...
                    T, obj.CriticalTemperature);
            end
            Tr = obj.reducedTemperature(T);
            alpha = obj.temperatureCorrectionFactor(Tr);
            a = alpha*obj.AttractionParam;
            b = obj.RepulsionParam;
            x = roots(obj.dPdVPolyEq(T,a,b));
            V = x(imag(x) == 0);
            V = V(V > b);
            V = sort(V);
            P = obj.pressure(T,V);
        end
    end
end